common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -795,13 +801,41 @@ public void registerExecutor(String appId, ExecutorShuffleInfo executorInfo) {
   }
 
   /**
-   * Close the DB during shutdown
+   * Shutdown mergedShuffleCleaner and close the DB during shutdown
    */
   @Override
   public void close() {
+    if (!mergedShuffleCleaner.isShutdown()) {
+      // SPARK-40186：Use two phases shutdown refer to
+      // https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ExecutorService.html
+      // Use two phases shutdown can prevent new tasks and wait for executing tasks to
+      // complete gracefully, and once timeout is reached, we want to interrupt running tasks,
+      // so that they fail. This is to prevent updates to shuffle state db after it is closed.
+      try {
+        mergedShuffleCleaner.shutdown();
+        // Wait a while for existing tasks to terminate
+        if (!mergedShuffleCleaner.awaitTermination(cleanerShutdownTimeout, TimeUnit.SECONDS)) {
+          List<Runnable> unfinishedTasks = mergedShuffleCleaner.shutdownNow();
+          logger.warn("There are still {} tasks not completed in mergedShuffleCleaner " +
+            "after {} seconds.", unfinishedTasks.size(), cleanerShutdownTimeout);
+          // Wait a while for tasks to respond to being cancelled
+          if (!mergedShuffleCleaner.awaitTermination(cleanerShutdownTimeout, TimeUnit.SECONDS)) {
+            logger.warn("mergedShuffleCleaner did not terminate");
+          }
+        }
+      } catch (InterruptedException ignored) {
+        // (Re-)Cancel if current thread also interrupted
+        List<Runnable> unfinishedTasks = mergedShuffleCleaner.shutdownNow();
```

Comment:
Do this only if the interruption is for the first `awaitTermination`.
If we have already invoked `shutdownNow`, we can skip it

Reply:
skip Re-Cancel mergedShuffleCleaner?

Reply:
If the `awaitTermination` in line 817  responds to the 'InterruptedException',  I think `shutdownNow` should be called here, am I right?

Reply:
Yes, only for that case - there is a second `awaitTermination` (after `shutdownNow`), which can also be interrupted - for which we dont want the `shutdownNow` + log messages

Reply:
[80869e9](https://github.com/apache/spark/pull/37624/commits/80869e93f4ba89a888339e9484631f914f0da71b) remove it

Reply:
To clarify, I did not mean remove it entirely.
But rather, do it only if we have not yet done `shutdownNow` - that is, in case we got interrupted due to the first `awaitTermination`

Reply:
I was wrong